sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,984 @@
+
+        None and -1 means using all processors.
+    """
+
+    cdef:
+        readonly DatasetsPair datasets_pair
+
+        ITYPE_t n_threads
+        ITYPE_t effective_omp_n_thread
+        ITYPE_t n_samples_chunk, chunk_size
+
+        ITYPE_t n_samples_X, X_n_samples_chunk, X_n_chunks, X_n_samples_remainder
+        ITYPE_t n_samples_Y, Y_n_samples_chunk, Y_n_chunks, Y_n_samples_remainder
+
+    @classmethod
+    def valid_metrics(cls) -> List[str]:
+        excluded = {
+            "pyfunc",  # is relatively slow because we need to coerce data as np arrays
+            "mahalanobis", # is numerically unstable
+            # TODO: In order to support discrete distance metrics, we need to have a
+            # simultaneous sort which breaks ties on indices when distances are identical.
+            # The best might be using a std::sort and a Comparator which might need
+            # AoS instead of SoA (currently used).
+            "hamming",
+            *BOOL_METRICS,
+        }
+        return sorted({"fast_euclidean", "fast_sqeuclidean",
+                       *METRIC_MAPPING.keys()}.difference(excluded))
+
+    @classmethod
+    def is_usable_for(cls, X, Y, metric) -> bool:
+        """Return True if the PairwiseDistancesReduction can be used for the given parameters.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples_X, n_features)
+            Input data.
+
+        Y : {ndarray, sparse matrix} of shape (n_samples_Y, n_features)
+            Input data.
+
+        metric : str, default='euclidean'
+            The distance metric to use.
+            For a list of available metrics, see the documentation of
+            :class:`~sklearn.metrics.DistanceMetric`.
+
+        Returns
+        -------
+        True if the PairwiseDistancesReduction can be used, else False.
+        """
+        # Coercing to np.array to get the dtype
+        # TODO: what is the best way to get lists' dtype?
+        X = np.asarray(X) if not isinstance(X, (np.ndarray, scipy.sparse.spmatrix)) else X
+        Y = np.asarray(Y) if not isinstance(Y, (np.ndarray, scipy.sparse.spmatrix)) else Y
+        # TODO: support sparse arrays and 32 bits
+        return (not issparse(X) and X.dtype == np.float64 and X.ndim == 2 and
+                not issparse(Y) and Y.dtype == np.float64 and Y.ndim == 2 and
+                metric in cls.valid_metrics())
+
+    def __init__(
+        self,
+        DatasetsPair datasets_pair,
+        chunk_size=None,
+        n_threads=None,
+     ):
+        cdef:
+            ITYPE_t n_samples_chunk, X_n_full_chunks, Y_n_full_chunks
+
+        if chunk_size is None:
+            chunk_size = get_config().get("pairwise_dist_chunk_size", 256)
+
+        self.chunk_size = check_scalar(chunk_size, "chunk_size", Integral, min_val=1)
+
+        self.effective_omp_n_thread = _openmp_effective_n_threads(n_threads)
+
+        n_samples_chunk = max(20, chunk_size)
```

Comment:
Since `n_sample_chunk` is a max of two numbers, it may be different compare to `self.chunk_size`.

I think it would be okay to restrict `self.chunk_size` to be have `min_val=20` and use `self.chunk_size` directly.

Reply:
I am not sure I understand your suggestion here. How shall I use them separately (if this is what you propose)?

Reply:
If one sets the global config option `pairwise_dist_chunk_size=10`, the actual chunk size is 20. Concretely:

```python
self.chunk_size = check_scalar(chunk_size, "chunk_size", Integral, min_val=20)

# and use self.chunk_size directly
```

Reply:
I think @thomasjpfan mean that we could `raise ValueError` if `self.chunk_size < 20` (by setting `min_val=20` above) and use `chunk_size` directly here.